// debug/src/browser.js
/* eslint-env browser */

/**
 * This is the web browser implementation of `debug()`.
 */

exports.formatArgs = formatArgs;
exports.save = save;
exports.load = load;
exports.useColors = useColors;
exports.storage = localstorage();
exports.destroy = (() => {
	let warned = false;

	return () => {
		if (!warned) {
			warned = true;
			console.warn('Instance method `debug.destroy()` is deprecated and no longer does anything. It will be removed in the next major version of `debug`.');
		}
	};
})();

/**
 * Colors.
 */

exports.colors = [
	'***REMOVED***0000CC',
	'***REMOVED***0000FF',
	'***REMOVED***0033CC',
	'***REMOVED***0033FF',
	'***REMOVED***0066CC',
	'***REMOVED***0066FF',
	'***REMOVED***0099CC',
	'***REMOVED***0099FF',
	'***REMOVED***00CC00',
	'***REMOVED***00CC33',
	'***REMOVED***00CC66',
	'***REMOVED***00CC99',
	'***REMOVED***00CCCC',
	'***REMOVED***00CCFF',
	'***REMOVED***3300CC',
	'***REMOVED***3300FF',
	'***REMOVED***3333CC',
	'***REMOVED***3333FF',
	'***REMOVED***3366CC',
	'***REMOVED***3366FF',
	'***REMOVED***3399CC',
	'***REMOVED***3399FF',
	'***REMOVED***33CC00',
	'***REMOVED***33CC33',
	'***REMOVED***33CC66',
	'***REMOVED***33CC99',
	'***REMOVED***33CCCC',
	'***REMOVED***33CCFF',
	'***REMOVED***6600CC',
	'***REMOVED***6600FF',
	'***REMOVED***6633CC',
	'***REMOVED***6633FF',
	'***REMOVED***66CC00',
	'***REMOVED***66CC33',
	'***REMOVED***9900CC',
	'***REMOVED***9900FF',
	'***REMOVED***9933CC',
	'***REMOVED***9933FF',
	'***REMOVED***99CC00',
	'***REMOVED***99CC33',
	'***REMOVED***CC0000',
	'***REMOVED***CC0033',
	'***REMOVED***CC0066',
	'***REMOVED***CC0099',
	'***REMOVED***CC00CC',
	'***REMOVED***CC00FF',
	'***REMOVED***CC3300',
	'***REMOVED***CC3333',
	'***REMOVED***CC3366',
	'***REMOVED***CC3399',
	'***REMOVED***CC33CC',
	'***REMOVED***CC33FF',
	'***REMOVED***CC6600',
	'***REMOVED***CC6633',
	'***REMOVED***CC9900',
	'***REMOVED***CC9933',
	'***REMOVED***CCCC00',
	'***REMOVED***CCCC33',
	'***REMOVED***FF0000',
	'***REMOVED***FF0033',
	'***REMOVED***FF0066',
	'***REMOVED***FF0099',
	'***REMOVED***FF00CC',
	'***REMOVED***FF00FF',
	'***REMOVED***FF3300',
	'***REMOVED***FF3333',
	'***REMOVED***FF3366',
	'***REMOVED***FF3399',
	'***REMOVED***FF33CC',
	'***REMOVED***FF33FF',
	'***REMOVED***FF6600',
	'***REMOVED***FF6633',
	'***REMOVED***FF9900',
	'***REMOVED***FF9933',
	'***REMOVED***FFCC00',
	'***REMOVED***FFCC33'
];

/**
 * Currently only WebKit-based Web Inspectors, Firefox >= v31,
 * and the Firebug extension (any Firefox version) are known
 * to support "%c" CSS customizations.
 *
 * TODO: add a `localStorage` variable to explicitly enable/disable colors
 */

// eslint-disable-next-line complexity
function useColors() {
	// NB: In an Electron preload script, document will be defined but not fully
	// initialized. Since we know we're in Chrome, we'll just detect this case
	// explicitly
	if (typeof window !== 'undefined' && window.process && (window.process.type === 'renderer' || window.process.__nwjs)) {
		return true;
	}

	// Internet Explorer and Edge do not support colors.
	if (typeof navigator !== 'undefined' && navigator.userAgent && navigator.userAgent.toLowerCase().match(/(edge|trident)\/(\d+)/)) {
		return false;
	}

	// Is webkit? http://stackoverflow.com/a/16459606/376773
	// document is undefined in react-native: https://github.com/facebook/react-native/pull/1632
	return (typeof document !== 'undefined' && document.documentElement && document.documentElement.style && document.documentElement.style.WebkitAppearance) ||
		// Is firebug? http://stackoverflow.com/a/398120/376773
		(typeof window !== 'undefined' && window.console && (window.console.firebug || (window.console.exception && window.console.table))) ||
		// Is firefox >= v31?
		// https://developer.mozilla.org/en-US/docs/Tools/Web_Console***REMOVED***Styling_messages
		(typeof navigator !== 'undefined' && navigator.userAgent && navigator.userAgent.toLowerCase().match(/firefox\/(\d+)/) && parseInt(RegExp.$1, 10) >= 31) ||
		// Double check webkit in userAgent just in case we are in a worker
		(typeof navigator !== 'undefined' && navigator.userAgent && navigator.userAgent.toLowerCase().match(/applewebkit\/(\d+)/));
}

/**
 * Colorize log arguments if enabled.
 *
 * @api public
 */

function formatArgs(args) {
	args[0] = (this.useColors ? '%c' : '') +
		this.namespace +
		(this.useColors ? ' %c' : ' ') +
		args[0] +
		(this.useColors ? '%c ' : ' ') +
		'+' + module.exports.humanize(this.diff);

	if (!this.useColors) {
		return;
	}

	const c = 'color: ' + this.color;
	args.splice(1, 0, c, 'color: inherit');

	// The final "%c" is somewhat tricky, because there could be other
	// arguments passed either before or after the %c, so we need to
	// figure out the correct index to insert the CSS into
	let index = 0;
	let lastC = 0;
	args[0].replace(/%[a-zA-Z%]/g, match => {
		if (match === '%%') {
			return;
		}
		index++;
		if (match === '%c') {
			// We only are interested in the *last* %c
			// (the user may have provided their own)
			lastC = index;
		}
	});

	args.splice(lastC, 0, c);
}

/**
 * Invokes `console.debug()` when available.
 * No-op when `console.debug` is not a "function".
 * If `console.debug` is not available, falls back
 * to `console.log`.
 *
 * @api public
 */
exports.log = console.debug || console.log || (() => {});

/**
 * Save `namespaces`.
 *
 * @param {String} namespaces
 * @api private
 */
function save(namespaces) {
	try {
		if (namespaces) {
			exports.storage.setItem('debug', namespaces);
		} else {
			exports.storage.removeItem('debug');
		}
	} catch (error) {
		// Swallow
		// XXX (@Qix-) should we be logging these?
	}
}

/**
 * Load `namespaces`.
 *
 * @return {String} returns the previously persisted debug modes
 * @api private
 */
function load() {
	let r;
	try {
		r = exports.storage.getItem('debug');
	} catch (error) {
		// Swallow
		// XXX (@Qix-) should we be logging these?
	}

	// If debug isn't set in LS, and we're in Electron, try to load $DEBUG
	if (!r && typeof process !== 'undefined' && 'env' in process) {
		r = process.env.DEBUG;
	}

	return r;
}

/**
 * Localstorage attempts to return the localstorage.
 *
 * This is necessary because safari throws
 * when a user disables cookies/localstorage
 * and you attempt to access it.
 *
 * @return {LocalStorage}
 * @api private
 */

function localstorage() {
	try {
		// TVMLKit (Apple TV JS Runtime) does not have a window object, just localStorage in the global context
		// The Browser also has localStorage in the global context.
		return localStorage;
	} catch (error) {
		// Swallow
		// XXX (@Qix-) should we be logging these?
	}
}

module.exports = require('./common')(exports);

const {formatters} = module.exports;

/**
 * Map %j to `JSON.stringify()`, since no Web Inspectors do that by default.
 */

formatters.j = function (v) {
	try {
		return JSON.stringify(v);
	} catch (error) {
		return '[UnexpectedJSONParseError]: ' + error.message;
	}
};
