// mongoose/lib/helpers/symbols.js
'use strict';

exports.arrayAtomicsBackupSymbol = Symbol('mongoose***REMOVED***Array***REMOVED***atomicsBackup');
exports.arrayAtomicsSymbol = Symbol('mongoose***REMOVED***Array***REMOVED***_atomics');
exports.arrayParentSymbol = Symbol('mongoose***REMOVED***Array***REMOVED***_parent');
exports.arrayPathSymbol = Symbol('mongoose***REMOVED***Array***REMOVED***_path');
exports.arraySchemaSymbol = Symbol('mongoose***REMOVED***Array***REMOVED***_schema');
exports.documentArrayParent = Symbol('mongoose:documentArrayParent');
exports.documentIsSelected = Symbol('mongoose***REMOVED***Document***REMOVED***isSelected');
exports.documentIsModified = Symbol('mongoose***REMOVED***Document***REMOVED***isModified');
exports.documentModifiedPaths = Symbol('mongoose***REMOVED***Document***REMOVED***modifiedPaths');
exports.documentSchemaSymbol = Symbol('mongoose***REMOVED***Document***REMOVED***schema');
exports.getSymbol = Symbol('mongoose***REMOVED***Document***REMOVED***get');
exports.modelSymbol = Symbol('mongoose***REMOVED***Model');
exports.objectIdSymbol = Symbol('mongoose***REMOVED***ObjectId');
exports.populateModelSymbol = Symbol('mongoose.PopulateOptions***REMOVED***Model');
exports.schemaTypeSymbol = Symbol('mongoose***REMOVED***schemaType');
exports.sessionNewDocuments = Symbol('mongoose:ClientSession***REMOVED***newDocuments');
exports.scopeSymbol = Symbol('mongoose***REMOVED***Document***REMOVED***scope');
exports.validatorErrorSymbol = Symbol('mongoose:validatorError');
